File: 1576-reorder-routes-to-make-all-paths-lead-to-the-city-zero/reorder-routes-to-make-all-paths-lead-to-the-city-zero.java
class Solution {
    public int minReorder(int n, int[][] connections) {
        List<int[]>[] adj = new List[n];
        for (int i = 0; i < n; i++) adj[i] = new ArrayList<>();
        
        for (int[] c : connections) {
            adj[c[0]].add(new int[]{c[1], 1});
            adj[c[1]].add(new int[]{c[0], 0});
        }

        boolean[] inFamily = new boolean[n];
        int count = 0;
        
        List<Integer> familyList = new ArrayList<>();
        familyList.add(0);
        inFamily[0] = true;

        for (int i = 0; i < familyList.size(); i++) {
            int currentMember = familyList.get(i);
            
            for (int[] edge : adj[currentMember]) {
                int neighbor = edge[0];
                int isOutbound = edge[1];
                
                if (!inFamily[neighbor]) {
                    count += isOutbound;
                    
                    inFamily[neighbor] = true;
                    familyList.add(neighbor);
                }
            }
        }
        
        return count;
    }
}

File: 1576-reorder-routes-to-make-all-paths-lead-to-the-city-zero/reorder-routes-to-make-all-paths-lead-to-the-city-zero.cpp
class Solution {
public:
    int count = 0;
    vector<bool> visited;

    void dfs(int u, vector<vector<int>>& adj_out, vector<vector<int>>& adj_in) {
        visited[u] = true;

        for (int v : adj_out[u]) {
            if (!visited[v]) {
                count++; 
                dfs(v, adj_out, adj_in);
            }
        }

        for (int v : adj_in[u]) {
            if (!visited[v]) {
                dfs(v, adj_out, adj_in);
            }
        }
    }

    int minReorder(int n, vector<vector<int>>& connections) {
        vector<vector<int>> adj_out(n);
        vector<vector<int>> adj_in(n); 
        
        for (auto& c : connections) {
            adj_out[c[0]].push_back(c[1]);
            adj_in[c[1]].push_back(c[0]);
        }

        visited.assign(n, false);
        count = 0;
        dfs(0, adj_out, adj_in);
        
        return count;
    }
};